akka-cluster-tools/src/main/scala/akka/cluster/client/ClusterClient.scala

```diff
@@ -703,7 +908,45 @@ final class ClusterReceptionist(pubSubMediator: ActorRef, settings: ClusterRecep
       }
 
     case _: MemberEvent ⇒ // not of interest
+
+    case GetClusterClients ⇒
+      sender() ! ClusterClients(clientInteractions.keySet)
+
+    case CheckDeadlines ⇒
+      clientInteractions = clientInteractions.filter {
+        case (_, failureDetector) ⇒
+          failureDetector.isAvailable
+      }
+      publishClientsUnreachable()
+  }
+
+  def updateClientInteractions(client: ActorRef): Unit =
+    clientInteractions.get(client) match {
+      case Some(failureDetector) ⇒
+        failureDetector.heartbeat()
+      case None ⇒
+        val failureDetector = new DeadlineFailureDetector(acceptableHeartbeatPause, heartbeatInterval)
+        clientInteractions = clientInteractions + (client -> failureDetector)
+    }
+
+  def publishClientsUp(): Unit = {
+    val publishableClients = clientInteractions.keySet.to[HashSet]
+    (publishableClients -- clientsPublished)
+      .foreach { c ⇒
+        log.info("Received new contact from [{}]", c.path)
+        context.system.eventStream.publish(ClusterClientUp(c, self))
+      }
+    clientsPublished = publishableClients
   }
 
+  def publishClientsUnreachable(): Unit = {
+    val publishableClients = clientInteractions.keySet.to[HashSet]
+    (clientsPublished -- publishableClients)
```

Comment:
OK then.

> On 9 May 2016, at 6:25 PM, Patrik Nordwall <notifications@github.com> wrote:
> 
> In akka-cluster-tools/src/main/scala/akka/cluster/client/ClusterClient.scala <https://github.com/akka/akka/pull/20462#discussion_r62463674>:
> 
> >    }
> >  
> > +  def publishClientsUnreachable(): Unit = {
> > +    val publishableClients = clientInteractions.keySet.to[HashSet]
> > +    (clientsPublished -- publishableClients)
> I think diff would create more objects and I can't see that diff could implement itself without doing similar lookup
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub <https://github.com/akka/akka/pull/20462/files/22b7e59e609b89af8a35659692e7f8612bb66c33#r62463674>